### DISCUSSION TITLE
Here's some free Evennia advice.

### DISCUSSION THREAD

**Discussion by kovitikus** [evennia/evennia#2282] [General Discussion]
#### Reference Evennia's Source Code
First and foremost, you must read while you write. By this I mean that you need to reference the source code and the documentation while you are writing systems, until you become familiar enough to have things memorized. Even then, you'll still generally find yourself searching for things. It's a basic requirement when it comes to coding. Google search is your friend, but don't forget to use the GitHub search to find things in Evennia itself.

***

#### Read Documentation
Speaking of reading, I always tell people that the best thing you can do for yourself is to passively read everything you can find about Evennia. While you have downtime in real-life, instead of going to a social media site and wasting time, pull up some documentation on Evennia and passively read through it. Even if it's something you don't understand or aren't using immediately, that knowledge gets planted in your mind and you will be more apt to find it again and use it later.

***

#### Keep Moving
Learning Evennia and Python is all about consistency and momentum. Don't overwhelm yourself, but also don't get stuck taking too much of a break from development; if you can help it. It may be frustrating to start from nothing and have no clue what you are doing, but you will gain bits of success each time you sit down to complete a task.

***

#### Avoid Over-Engineering
If your code works, only touch it again when you need to make changes or you have an epiphany of how it could be structured better, to help you and others read it later down the road. This same advice also holds true for planning your game. It's not a bad idea to jot down your ideas, and even fully realize them in your spare time, but don't get caught up in creating your gameworld before you have any systems in place.

I can tell you from my own personal development experience that having too many aspects of a concept will serve to confuse you. What I mean by this is sheer raw data. One time I made a really big dictionary for my skill system. I started to plot out all of the skills and values I thought it might need. As time went on and I began really ironing out the details of the underlying systems of code, I realized that I had to make changes to the dictionary. It became a huge waste of time to sit there and edit every entry of the dictionary, because these changes happened several more times. I eventually cut down my dictionary so that I have enough information in there to test my systems with and nothing more.

***

#### Minimum Viable Product
Start your development with the most foundational mechanics that define your game. How will your game handle conveying a character moving from one room to the next? In what way do you want the appearance of a room to be returned from the room object?

Focus on getting your game to the most basic playable level that you can. This does not require you to logically design systems in the order of the player's experience. You do NOT have to start by learning EvMenu and designing a character generation system. So long as your game objects are created with the proper typeclass and attributes/tags/locks, you can test your systems with the most generic of objects.

**Comment by magnumopus** [comment]
Relevant to the first point of **Reference Evennia's Source Code**, I would like to suggest that a great place to start is [`DefaultObject`](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L193) and then it's children ([`DefaultCharacter`](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L2026), [`DefaultRoom`](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L2243), etc). These contain some methods like [`return_appearance()`](https://github.com/evennia/evennia/blob/master/evennia/objects/objects.py#L1650) that are very helpful to understand.

[`MuxCommand`](https://github.com/evennia/evennia/blob/master/evennia/commands/default/muxcommand.py#L13 )is great place to go from there.

**Reply by kovitikus** [reply]
`return_appearance()` is one of the first pieces of source code I read through and it helped me understand a lot about getting objects in the room and filtering through them! It was also the catalyst that propelled me forward into a momentum where I was able to read source code and learn on my own, outside of tutorials. I was really joyous after understanding how `return_appearance()` works that I delved deep on exploring other parts of the source code.

DefaultObject is really important as well, because it contains all of the most important methods you'll use when dealing with objects, such as `object.contents`, etc.

**Comment by Griatch** [comment]
Nice suggestions, @kovitikus! Helpful for newcomers for sure. :smiley: 

I should probably also point out the [Planning your game](https://www.evennia.com/docs/latest/Game-Planning.html) page from the Evennia documentation.

**Reply by kovitikus** [reply]
Yeah, some of the advice I've offered is a bit of a repeat of some sections of the `Planning your game` article. It's a great resource for someone to read when they first start out, for sure.
